odoo/addons/base/tests/test_ir_mail_server_smtpd.py

```diff
@@ -0,0 +1,443 @@
+        tempdir = tempfile.TemporaryDirectory(prefix='odoo-test-smtpd-')
+        cls.addClassCleanup(tempdir.cleanup)
+        cls.ssl_ca, cls.ssl_client, cls.ssl_server, cls.ssl_self_signed = \
+            _generate_certificates(Path(tempdir.name))
+
+        class TEST_SMTP(smtplib.SMTP):
+            def starttls(self, *, context):
+                if context is None:
+                    context = ssl._create_stdlib_context()  # what SMTP_SSL does
+                    # context = ssl.create_default_context()  # what it should do
+                context.load_verify_locations(cafile=str(cls.ssl_ca.cert))
+                super().starttls(context=context)
+        patcher = patch('smtplib.SMTP', TEST_SMTP)
+        patcher.start()
+        cls.addClassCleanup(patcher.stop)
+
+        class TEST_SMTP_SSL(smtplib.SMTP_SSL):
+            def _get_socket(self, *args, **kwargs):
+                # self.context = ssl.create_default_context()  # what it should do
+                self.context.load_verify_locations(cafile=str(cls.ssl_ca.cert))
+                return super()._get_socket(*args, **kwargs)
+        patcher = patch('smtplib.SMTP_SSL', TEST_SMTP_SSL)
+        patcher.start()
+        cls.addClassCleanup(patcher.stop)
+
+        # reactivate sending emails during this test suite, make sure
+        # NOT TO send emails using another ir.mail_server than the one
+        # created in setUp!
+        patcher = patch.object(cls.registry['ir.mail_server'], '_is_test_mode')
+        mock = patcher.start()
+        mock.return_value = False
+        cls.addClassCleanup(patcher.stop)
+
+        # fix runbot, docker uses a single ipv4 stack but it gives ::1
+        # when resolving "localhost" (so stupid), use the following to
+        # force aiosmtpd/odoo to bind/connect to a fixed ipv4 OR ipv6
+        # address.
+        family, _, cls.port = _find_free_local_address()
+        cls.localhost = getaddrinfo('localhost', cls.port, family)
+        cls.startClassPatcher(patch('socket.getaddrinfo', cls.getaddrinfo))
+
+    @classmethod
+    def getaddrinfo(cls, host, port, *args, **kwargs):
+        """
+        Resolve both "localhost" and "notlocalhost" on the ip address
+        bound by aiosmtpd inside `start_smtpd`.
+        """
+        if host in ('localhost', 'notlocalhost') and port == cls.port:
+            return cls.localhost
+        return getaddrinfo(host, port, family=0, type=0, proto=0, flags=0)
+
+    def setUp(self):
+        super().setUp()
+        # Add an outgoing mail server for the server we start with ``start_smtpd``
+        self.mail_server = self.env['ir.mail_server'].create({
+            'name': 'test smtpd',
+            'from_filter': 'localhost',
+            'smtp_host': 'localhost',
+            'smtp_port': self.port,
+            'smtp_authentication': 'login',
+            'smtp_user': '',
+            'smtp_pass': '',
+        })
+
+    @contextlib.contextmanager
+    def start_smtpd(
```

Comment:
:socks: Strange writing, takes a lot of vertical space for nothing ? Remember we don't enforce line length at odoo, especially when it creates code like that (being smart > linter).

Reply:
I enforce myself a 96 characters limit. 96 characters is the size of the github editor, in review, in split mode, with the file tree open, on a 1920x1080 screen, with the default zoom, with the default font on our linux laptop. It makes that one line of code = one line on the github editor when doing reviews. Yes I'm crazy but you already knew that.

And I genuinely prefer when functions with many arguments are written vertically, but that's my own preference.

Went for:

```py
    def start_smtpd(
        self, encryption, ssl_context=None, auth_required=True, stop_on_cleanup=True
    ):
```
which both fit on 96 characters and is better according to your preferences